tokio-boring/src/bridge.rs

```diff
@@ -0,0 +1,86 @@
+//! Bridge between sync IO traits and async tokio IO traits.
+
+use std::fmt;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll, Waker};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+pub(crate) struct AsyncStreamBridge<S> {
+    pub(crate) stream: S,
+    waker: Option<Waker>,
+}
+
+impl<S> AsyncStreamBridge<S> {
+    pub(crate) fn new(stream: S) -> Self
+    where
+        S: AsyncRead + AsyncWrite + Unpin,
+    {
+        Self {
+            stream,
+            waker: None,
+        }
+    }
+
+    pub(crate) fn set_waker(&mut self, ctx: Option<&mut Context<'_>>) {
+        self.waker = ctx.map(|ctx| ctx.waker().clone())
+    }
+
+    /// # Panics
+    ///
+    /// Panics if the bridge has no waker.
+    pub(crate) fn with_context<F, R>(&mut self, f: F) -> R
+    where
+        S: Unpin,
+        F: FnOnce(&mut Context<'_>, Pin<&mut S>) -> R,
+    {
+        let mut ctx =
+            Context::from_waker(self.waker.as_ref().expect("missing task context pointer"));
```

Comment:
I would be very confused if I saw my program panic with this error, and would immediately go on completely the wrong path to debugging the issue.

Reply:
Can you suggest an alternative e panic message?

Reply:
"Internal error: missing waker in bridge"?

Reply:
I suppose this is a bit of a nit.

Reply:
I changed the message, using prefix "BUG:" like for all the other similar panic messages that exist in those crates.